### PR TITLE
Add mochajs framework for automated UI and notebook tests

### PR DESCRIPTION
**Pull request by yebrahim**
This PR moves tests to Javascript and adds UI regression tests.

It uses a standalone headless Chrome browser in a docker container provided by selenium, and nodejs bindings for selenium to write tests that launch a browser instance, take screenshots, and execute notebooks. Using the docker container reduces cross platform dependencies, provides a stable platform for tests, and reduces the overall test time, since no image needs to be built like before.

For UI tests, I use the resemblejs package to compare screenshots against saved (golden) images, which is able to tolerate small anti-aliasing differences sometimes caused by selenium webdrivers.

For notebook tests, I rewrote the python code that started the browser in Javascript. This way the code is smaller, and can be run using the same test runner.

Tests were written as a nodejs package, so in order to run them locally, you can do `npm install` then `npm test`. The one dependency is you have to be able to install Cairo, which is a resemblejs dependency, on your system. Take a look at the Travis config changes.

### CODE REVIEW

File: sources/web/datalab/static/appbar.html [hunk 1]
```diff
@@ -12,12 +12,12 @@
     <span class="autosave_status"></span>
   </span>
   <div class="btn-toolbar pull-right right-toolbar">
-    <div class="btn-group">
-      <button id="ungitButton" title="Open ungit on the notebooks repository" class="toolbar-btn">
+    <div id="ungitButton" class="btn-group">
```
Comment (jimmc) [inline sources/web/datalab/static/appbar.html]: I've seen people use divs as buttons, but I don't recall seeing a div used as a button, with a button element inside it. This  seems a bit confusing to me. Why do you need to move the click handler to the div? Perhaps that deserves a comment.

File: test/notebook/test.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      until = selenium.until;
+const test = require('selenium-webdriver/testing');
+const fs = require('fs');
+
+var driver = null;
+
+const notebooks_config = JSON.parse(fs.readFileSync('notebook/spec.json'))
+const timeOut = 60000;
+
+test.before(function() {
```
Comment (jimmc) [inline test/notebook/test.js]: Should we start using arrow syntax for anonymous functions?

File: test/ui/test.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      until = selenium.until;
+const test = require('selenium-webdriver/testing');
+const fs = require('fs');
+const resemble = require('node-resemble');
+
+var driver = null;
+
+const suiteTimeout = 60000;
+const misMatchThreshold = 0;
+const goldenPathPrefix = 'ui/goldens/';
```
Comment (jimmc) [inline test/ui/test.js]: I think I would prefer no 's' on this name. That also makes it consistent with the 'broken' directory name.
Making them consistent by naming the other one 'brokens' seems a lot worse.

File: test/ui/test.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        fs.writeFileSync(brokenPathPrefix + testName + '.png', shot, 'base64');
+      }
+
+      assert(data.misMatchPercentage <= misMatchThreshold,
+             'Images for test ' + testName + ' are different');
+    });
+  });
+}
+
+test.describe('UI tests', function() {
```
Comment (jimmc) [inline test/ui/test.js]: This string should be the name of the component being tested in this section, and the 'it' string should be a verb phrase that completes a sentence starting with the 'describe' string. I suggest you break this describe into two calls the describe: the first one would be test.descibe('Main window'), and the second one would be test.describe('Help window').

File: test/ui/test.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+  test.it('Makes sure window looks fine at the beginning of tests', function() {
```
Comment (jimmc) [inline test/ui/test.js]: Given my comment about 'describe' above, this string would be something like 'appears correctly before any actions have been taken'.

File: test/ui/test.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    return screenshotAndCompare('body.png', 'body');
+  });
+
+  test.it('Opens help menu and makes sure it looks fine', function() {
+    return driver.findElement(By.id('helpButton'))
+    .then(function(button) {
```
Comment (jimmc) [inline test/ui/test.js]: I think this would look better as a one-line arrow function.

File: test/ui/test.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  });
+}
+
+test.describe('UI tests', function() {
+
+  test.it('Makes sure window looks fine at the beginning of tests', function() {
+    return screenshotAndCompare('body.png', 'body');
+  });
+
+  test.it('Opens help menu and makes sure it looks fine', function() {
```
Comment (jimmc) [inline test/ui/test.js]: This string should be something like 'opens when Help button is clicked'.

File: test/ui/test.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    return driver.findElement(By.id('helpButton'))
+    .then(function(button) {
+      button.click();
+    })
+    .then(screenshotAndCompare('bodyWithHelp.png', 'bodyWithHelp'));
+  });
+
+  test.it('Close help menu and makes sure body looks fine', function() {
```
Comment (jimmc) [inline test/ui/test.js]: This should be something like 'closes when body is clicked'.

File: test/README.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
+# Testing Google Cloud DataLab
+
+This directory contains tests Datalab. Tests are all written in Javascript, and they use a  to
```
Comment (jimmc) [inline test/README.md]: Looks like something is missing here: "they use a  to"

File: test/ui/test.js (hunk 4 above)
Comment (jimmc) [inline test/ui/test.js]: Looks good, thanks.

File: sources/web/datalab/static/appbar.html (hunk 1 above)
Comment (jimmc) [inline sources/web/datalab/static/appbar.html]: Here is a SO article with some discussion that mentions accessibility and screen readers as reasons not to us a div as a button: http://stackoverflow.com/questions/2777256/making-buttons-button-or-div

Comment (yebrahim) [inline sources/web/datalab/static/appbar.html]: Synced with @jimmc on this. We might want to pick up some cleanup work later to make the code consistent.

**Comment by yebrahim** [comment]
Oh I missed that. Fixed.

**Pull request merged**